### PR TITLE
Refactor `checkResult()`

### PR DESCRIPTION
**Pull request by abdulmth**

### CODE REVIEW

File: packages/sdk-js/src/DidHelpers/checkResult.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+function checkStatus(result: SubmittableResultValue): {
+  status: TransactionResult['status'] | undefined
+  error: Error | undefined
+  blockHash: HexString | undefined
+  blockNumber: BigInt | undefined
+} {
+  let status: TransactionResult['status'] | undefined
+  let blockHash
+  let blockNumber
+  let error
```
Comment (rflechtner) [inline packages/sdk-js/src/DidHelpers/checkResult.ts:58]: let's use returns inside of the switch instead

File: packages/sdk-js/src/DidHelpers/checkResult.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+import type { Did, HexString } from '@kiltprotocol/types'
+import type { SharedArguments, TransactionResult } from './interfaces.js'
+import { assertStatus, mapError } from './common.js'
+
+function checkEventsForErrors(
+  api: ApiPromise,
+  txEvents: EventRecord[] = []
+): Error | undefined {
+  let error
+  txEvents.forEach(({ event }) => {
```
Comment (rflechtner) [inline packages/sdk-js/src/DidHelpers/checkResult.ts:26]:
```suggestion
  return txEvents.find(({ event }) => {
```
combine this with returns (I'll make one example below) and our function becomes simpler and faster (exits on first error)

File: packages/sdk-js/src/DidHelpers/checkResult.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    if (api.events.system.ExtrinsicFailed.is(event)) {
+      error = mapError(event.data[0], api)
```
Comment (rflechtner) [inline packages/sdk-js/src/DidHelpers/checkResult.ts:28]:
```suggestion
      return mapError(event.data[0], api)
```

Comment (rflechtner) [reply packages/sdk-js/src/DidHelpers/checkResult.ts:28]: same for all other branches

File: packages/sdk-js/src/DidHelpers/checkResult.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    const eventError = checkEventsForErrors(api, txEvents)
+    if (eventError !== undefined) {
+      error = eventError
+    }
+
+    const eventMatch = checkIfEventsMatch(expectedEvents, txEvents)
+    const isSuccess = !error && eventMatch
+
+    status = isSuccess ? 'confirmed' : 'failed'
+    if (!isSuccess && !error) {
```
Comment (rflechtner) [inline packages/sdk-js/src/DidHelpers/checkResult.ts:179]:
```suggestion
    if (!isSuccess && !eventMatch) {
```
why not be more explicit?

File: packages/sdk-js/src/DidHelpers/common.ts [hunk 5]
```diff
@@ -9,38 +9,25 @@
...
 
 import type {
   AcceptedPublicKeyEncodings,
   SharedArguments,
   TransactionHandlers,
-  TransactionResult,
 } from './interfaces.js'
 
-function mapError(err: SpRuntimeDispatchError, api: ApiPromise): Error {
+export function mapError(err: SpRuntimeDispatchError, api: ApiPromise): Error {
```
Comment (rflechtner) [inline packages/sdk-js/src/DidHelpers/common.ts:22]: this one and the next are only used in checkResult rn, so they could move there too

File: packages/sdk-js/src/DidHelpers/checkResult.ts (hunk 4 above)
Comment (abdulmth) [reply packages/sdk-js/src/DidHelpers/checkResult.ts]: 5fa8bed

File: packages/sdk-js/src/DidHelpers/checkResult.ts (hunk 3 above)
Comment (rflechtner) [reply packages/sdk-js/src/DidHelpers/checkResult.ts:47]: ah yes of course, didn't think it through enough. Yes, that'll do the trick

**Review by rflechtner — APPROVED** [review]
Nice!

**Pull request merged**